### PR TITLE
api:add api for history_hot_region info;cluster add storage for hot region;config add config for hot-region storage

### PR DESCRIPTION
**Pull request by qidi1**
Signed-off-by: qidi1 1083369179@qq.com

### What problem does this PR solve?
https://github.com/tikv/pd/pull/4019
https://github.com/tikv/pd/pull/4020
https://github.com/tikv/pd/pull/4021

Proposal: [pingcap/tidb#27487](https://github.com/pingcap/tidb/pull/27487)
Associated PR:[tikv/tidb#27224](https://github.com/pingcap/tidb/pull/27224)
Associated Issue: [pingcap/tidb#25281](https://github.com/pingcap/tidb/issues/25281)
[pingcap/tidb#27373](https://github.com/pingcap/tidb/pull/27373)
[pingcap/tidb#27374](https://github.com/pingcap/tidb/pull/27374)

### What is changed and how it works?
Store hot-region information regularly,Provide  api interface for requesting historical hot-region information
### Check List

Tests

- Unit test

Code changes

- Has configuration change
 add hot-regions-reserved-days and hot-regions-write-interval in scheduler,hot-regions-reserved-days means how many days of information are kept,hot-regions-write-interval means the time interval for pulling information
- Has HTTP API interfaces change (Don't forget to [add the declarative for API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
Add post api /hotspot/regions/history
- Has persistent data change
store hot region info

### Side effects
- Possible performance regression
store hot region data may have some performance to pd,under 1000 regions,10 min write intervals  context,to read all data in a month needs 18s, 1.7s is needed for every month  Leveldb compaction

### Related changes

Associated PR: https://github.com/pingcap/tidb/pull/27224

### Release note

```release-note
Store hot-region information regularly,Provide  api interface for requesting historical hot-region information
```

### CODE REVIEW

File: server/cluster/hot_region_storage.go [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+	"github.com/tikv/pd/pkg/encryption"
+	"github.com/tikv/pd/pkg/errs"
+	"github.com/tikv/pd/server/encryptionkm"
+	"github.com/tikv/pd/server/kv"
+	"github.com/tikv/pd/server/member"
+	"github.com/tikv/pd/server/statistics"
+)
+
+// HotRegionStorage used to storage hot region info,
+// It will pull the hot region information according to the pullInterval interval
```
Comment (IcePigZDB) [inline server/cluster/hot_region_storage.go:26]: is used to

File: server/cluster/hot_region_storage.go [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+		cluster:              cluster,
+		member:               member,
+	}
+	h.backgroundFlush()
+	h.backgroundDelete()
+	return &h, nil
+}
+
+//delete hot_region info which update_time is smaller than time.Now() minus /remain day in the background
+func (h *HotRegionStorage) backgroundDelete() {
```
Comment (IcePigZDB) [inline server/cluster/hot_region_storage.go:86]: add space.

File: server/cluster/hot_region_storage.go [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+				h.delete()
+			case <-h.hotRegionInfoCtx.Done():
+				return
+			}
+		}
+	}()
+
+}
+
+//write hot_region info into db in the background
```
Comment (IcePigZDB) [inline server/cluster/hot_region_storage.go]: add space

File: server/cluster/hot_region_storage.go [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+					}
+				}
+			case <-h.hotRegionInfoCtx.Done():
+				return
+			}
+		}
+	}()
+}
+
+//NewIterator return a iterator which can traverse all data as reqeust
```
Comment (IcePigZDB) [inline server/cluster/hot_region_storage.go]: add space.

File: server/cluster/hot_region_storage.go [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+		return nil, err
+	}
+	message.StartKey = region.StartKey
+	message.EndKey = region.EndKey
+	message.EncryptionMeta = nil
+	return &message, nil
+}
+
+//HotRegionStorePath generate hot region store key for HotRegionStorage
+//TODO:find a better place to put this function
```
Comment (IcePigZDB) [inline server/cluster/hot_region_storage.go]: better place

File: server/cluster/hot_region_storage.go (hunk 4 above)
Comment (IcePigZDB) [reply server/cluster/hot_region_storage.go]: Would it make more sense to put this NewIteratorbelow below Iterator struct？

File: server/statistics/history_hot_region.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+	// Encryption metadata for start_key and end_key. encryption_meta.iv is IV for start_key.
+	// IV for end_key is calculated from (encryption_meta.iv + len(start_key)).
+	// The field is only used by PD and should be ignored otherwise.
+	// If encryption_meta is empty (i.e. nil), it means start_key and end_key are unencrypted.
+	EncryptionMeta *encryptionpb.EncryptionMeta `json:"encryption_meta,omitempty"`
+}
+
+// HistoryHotRegionsRequest wrap request condition from tidb.
+//it is request from tidb
+//TODO
```
Comment (IcePigZDB) [inline server/statistics/history_hot_region.go]: better place

File: server/statistics/history_hot_region.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+	EndKey        []byte  `json:"end_key,omitempty"`
+	// Encryption metadata for start_key and end_key. encryption_meta.iv is IV for start_key.
+	// IV for end_key is calculated from (encryption_meta.iv + len(start_key)).
+	// The field is only used by PD and should be ignored otherwise.
+	// If encryption_meta is empty (i.e. nil), it means start_key and end_key are unencrypted.
+	EncryptionMeta *encryptionpb.EncryptionMeta `json:"encryption_meta,omitempty"`
+}
+
+// HistoryHotRegionsRequest wrap request condition from tidb.
+//it is request from tidb
```
Comment (IcePigZDB) [inline server/statistics/history_hot_region.go]: add space.

File: server/server.go [hunk 8]
```diff
@@ -391,7 +393,15 @@ func (s *Server) startServer(ctx context.Context) error {
 	s.basicCluster = core.NewBasicCluster()
 	s.cluster = cluster.NewRaftCluster(ctx, s.GetClusterRootPath(), s.clusterID, syncer.NewRegionSyncer(s), s.client, s.httpClient)
 	s.hbStreams = hbstream.NewHeartbeatStreams(ctx, s.clusterID, s.cluster)
-
+	//initial hot_region_storage in here
```
Comment (IcePigZDB) [inline server/server.go:395]: add space and remove `in here`

File: server/server.go [hunk 9]
```diff
@@ -143,6 +143,8 @@ type Server struct {
 
 	// Store as map[string]*grpc.ClientConn
 	clientConns sync.Map
+	//hot region history info storeage
```
Comment (IcePigZDB) [inline server/server.go:145]: add space

File: server/cluster/hot_region_storage_test.go [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+//Print dir size
+func PrintDirSize(path string) {
+	size, err := DirSizeB(path)
+	if err != nil {
+		log.Fatal(err)
+	}
+	fmt.Printf("file size %d\n", size)
+}
+
+//getFileSize get file size by path(B)
```
Comment (IcePigZDB) [inline server/cluster/hot_region_storage_test.go:385]: add space

File: server/cluster/hot_region_storage_test.go [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+			StartKey:    []byte(fmt.Sprintf("%020d", i)),
+			EndKey:      []byte(fmt.Sprintf("%020d", i+1)),
+			RegionEpoch: &metapb.RegionEpoch{ConfVer: 2, Version: 2},
+		}
+		regions = append(regions, core.NewRegionInfo(region, peers[0]))
+	}
+	return regions
+}
+
+//Print dir size
```
Comment (IcePigZDB) [inline server/cluster/hot_region_storage_test.go:376]: add space

File: server/cluster/hot_region_storage_test.go [hunk 12]
```diff
@@ -0,0 +1,395 @@
...
+		peers := region.GetPeers()
+		peer := peers[rand.Intn(len(peers))]
+		if stats[peer.StoreId] == nil {
+			stats[peer.StoreId] = &statistics.HotPeersStat{}
+		}
+		statShow := statistics.HotPeerStatShow{
+			RegionID:       region.GetMeta().Id,
+			StoreID:        peer.StoreId,
+			LastUpdateTime: start,
+			HotDegree:      rand.Int(),
```
Comment (IcePigZDB) [inline server/cluster/hot_region_storage_test.go:339]: rand.Int() is very large, may be could %100

File: server/cluster/hot_region_storage_test.go [hunk 13]
```diff
@@ -0,0 +1,395 @@
...
+	hotRegionStorage *HotRegionStorage,
+	clear func(), err error) {
+	writePath := "./tmp"
+	ctx := context.Background()
+	_, opt, err := newTestScheduleConfig()
+	if err != nil {
+		return nil, nil, err
+	}
+	raft := newTestCluster(ctx, opt).RaftCluster
+	//delete data in between today and tomrrow
```
Comment (IcePigZDB) [inline server/cluster/hot_region_storage_test.go:270]: add space

File: server/cluster/hot_region_storage_test.go [hunk 14]
```diff
@@ -0,0 +1,395 @@
...
+		b.Fatal(err)
+	}
+	defer clear()
+	raft := regionStorage.cluster
+	endTime := time.Now()
+	regions := newTestHotRegions(1000, 3)
+	for _, region := range regions {
+		raft.putRegion(region)
+	}
+	//leveldb will compaction after 30 times delete
```
Comment (IcePigZDB) [inline server/cluster/hot_region_storage_test.go:197]: add space, latter the smae.

**Review by IcePigZDB — CHANGES_REQUESTED** [review]

File: conf/config.toml [hunk 15]
```diff
@@ -116,6 +116,10 @@
 ## When PD fails to receive the heartbeat from a store after the specified period of time,
 ## it adds replicas at other nodes.
 # max-store-down-time = "30m"
+## Controls the time interval between write hot regions info into leveldb
+# hot-regions-write-interval= "20m"
```
Comment (nolouch) [inline conf/config.toml:120]: I suggest 10m, and the default `hot-regions-reserved-days` 7 days may be enough. and could you add a comment about How much storage size will use with the mock?

**Comment by disksing** [comment]
There are enough reviewers so I'm unassigning myself :)

**Comment by qidi1** [comment]
If the write interval is 10 minutes; If only data is written, the data size of a month is about 550M. Both data is written and deleted in a month. The data size of a month is about 880M

File: conf/config.toml (hunk 15 above)
Comment (qidi1) [reply conf/config.toml:120]: If the write interval is 10 minutes; If only data is written, the data size of a month is about 550M. Both data is written and deleted in a month. The data size of a month is about 880M

Comment (qidi1) [reply conf/config.toml:120]: If the write interval is 10 minutes; If only data is written, the data size of seven days is about 141M. Both data is written and deleted in a month. The data size of a month is about 220M

File: server/handler.go [hunk 16]
```diff
@@ -903,6 +903,59 @@ func (h *Handler) SetStoreLimitTTL(data string, value float64, ttl time.Duration
...
+	} else {
+		hotRegionTypes = cluster.HotRegionTypes
+	}
+	iter := h.s.hotRegionStorage.NewIterator(hotRegionTypes, request.StartTime, request.EndTime)
+	results := make([]*statistics.HistoryHotRegion, 0)
+	regionSet := make(map[uint64]bool)
+	storeSet := make(map[uint64]bool)
+	peerSet := make(map[uint64]bool)
+	for _, id := range request.RegionIDs {
+		regionSet[id] = true
```
Comment (nolouch) [inline server/handler.go:920]: Why need request region IDs and peerIDs? I think only `StoresIds` and `KeyRange` are reasonable.

File: server/statistics/history_hot_region.go [hunk 17]
```diff
@@ -0,0 +1,50 @@
+package statistics
```
Comment (nolouch) [inline server/statistics/history_hot_region.go:1]: need to add license like other files.

File: server/cluster/hot_region_storage.go [hunk 18]
```diff
@@ -0,0 +1,320 @@
+package cluster
```
Comment (nolouch) [inline server/cluster/hot_region_storage.go:1]: license.

**Comment by nolouch** [comment]
BTW, I think your PR can split into three PRs:
- storage
- config
- API

more small PRs review more quickly. And the quality is more guaranteed.

**Comment by nolouch** [comment]
BTW, you can also supports pd-ctl command to check hot history region.

File: server/cluster/hot_region_storage.go [hunk 19]
```diff
@@ -0,0 +1,320 @@
...
+		value, err := json.Marshal(stat)
+		if err != nil {
+			return errs.ErrProtoMarshal.Wrap(err).GenWithStackByCause()
+		}
+		batch.Put([]byte(key), value)
+	}
+	if err := h.LeveldbKV.Write(batch, nil); err != nil {
+		return errs.ErrLevelDBWrite.Wrap(err).GenWithStackByCause()
+	}
+	h.batchHotInfo = make(map[string]*statistics.HistoryHotRegion)
```
Comment (Yisaer) [inline server/cluster/hot_region_storage.go:233]: What's this line used for?

**Review by Yisaer — COMMENTED** [review]
I recommended submitting single pr for storage part first

File: conf/simconfig.toml [hunk 20]
```diff
@@ -29,3 +29,6 @@ leader-schedule-limit = 32
 region-schedule-limit = 128
 replica-schedule-limit = 32
 merge-schedule-limit = 32
+## TODO:Find a better place to put this config
+hot-regions-reserved-days= "30"
+hot-regions-write-interval= "20m"
```
Comment (rleungx) [inline conf/simconfig.toml:34]: need an empty line

File: server/api/hot_status.go [hunk 21]
```diff
@@ -152,3 +154,30 @@ func (h *hotStatusHandler) GetHotStores(w http.ResponseWriter, r *http.Request)
 	}
 	h.rd.JSON(w, http.StatusOK, stats)
 }
+
+// @Tags hotspot
+// @Summary List the history hot regions.
+// @Accept json
+// @Produce json
+// @Success 200 {object} statistics.HistoryHotRegions
```
Comment (rleungx) [inline server/api/hot_status.go:162]: also need 500 here

File: server/cluster/hot_region_storage_test.go [hunk 22]
```diff
@@ -0,0 +1,395 @@
...
+	}
+	return
+}
+func writeIntoDB(regionStorage *HotRegionStorage,
+	regions []*core.RegionInfo, times int,
+	endTime time.Time) time.Time {
+	raft := regionStorage.cluster
+	for i := 0; i < times; i++ {
+		if i%1000 == 0 {
+			fmt.Println(i)
```
Comment (rleungx) [inline server/cluster/hot_region_storage_test.go:289]: use log?

File: server/statistics/history_hot_region.go [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+	// IV for end_key is calculated from (encryption_meta.iv + len(start_key)).
+	// The field is only used by PD and should be ignored otherwise.
+	// If encryption_meta is empty (i.e. nil), it means start_key and end_key are unencrypted.
+	EncryptionMeta *encryptionpb.EncryptionMeta `json:"encryption_meta,omitempty"`
+}
+
+// HistoryHotRegionsRequest wrap request condition from tidb.
+// it is request from tidb
+// TODO:find a better place to put this struct
+type HistoryHotRegionsRequest struct {
```
Comment (rleungx) [inline server/statistics/history_hot_region.go:35]: Should it be put into api package?

File: server/server.go [hunk 24]
```diff
@@ -390,7 +392,15 @@ func (s *Server) startServer(ctx context.Context) error {
 	s.basicCluster = core.NewBasicCluster()
 	s.cluster = cluster.NewRaftCluster(ctx, s.GetClusterRootPath(), s.clusterID, syncer.NewRegionSyncer(s), s.client, s.httpClient)
 	s.hbStreams = hbstream.NewHeartbeatStreams(ctx, s.clusterID, s.cluster)
-
+	//initial hot_region_storage in here.
+	hotRegionPath := filepath.Join(s.cfg.DataDir, "hot-region")
+	s.hotRegionStorage, err = cluster.NewHotRegionsStorage(
```
Comment (rleungx) [inline server/server.go:397]: How about moving to the same package to keep consistent with `NewRegionStorage`?

File: server/cluster/hot_region_storage.go [hunk 25]
```diff
@@ -0,0 +1,320 @@
...
+
+// Write hot_region info into db in the background.
+func (h *HotRegionStorage) backgroundFlush() {
+	ticker := time.NewTicker(h.pullInterval)
+	go func() {
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				if h.member.IsLeader() {
```
[the remark this answers is not in the document]
Comment (qidi1) [reply server/cluster/hot_region_storage.go:139]: Here, I want the delete function run  at a time point, such as one clock every day.

File: server/cluster/hot_region_storage.go [hunk 26]
```diff
@@ -0,0 +1,320 @@
...
+	h.backgroundDelete()
+	return &h, nil
+}
+
+// delete hot region which update_time is smaller than time.Now() minus remain day in the background.
+func (h *HotRegionStorage) backgroundDelete() {
+	// make delete happened in defaultDeleteTime clock.
+	now := time.Now()
+	next := now.Add(time.Hour * 24)
+	next = time.Date(next.Year(), next.Month(), next.Day(), defaultDeleteTime, 0, 0, 0, next.Location())
```
[the remark this answers is not in the document]
Comment (qidi1) [reply server/cluster/hot_region_storage.go:108]: Here, I want the delete function run at a time point, such as one clock every day.

File: server/cluster/hot_region_storage_test.go (hunk 22 above)
Comment (qidi1) [reply server/cluster/hot_region_storage_test.go:297]: This is the code that has not been deleted before. It has been deleted

**Comment by HunDunDM** [comment]
Maybe `/hotspot/history/regions` is more RESTful?

File: cmd/pd-server/docs/docs.go [hunk 27]
```diff
@@ -0,0 +1,74 @@
...
+	Version     string
+	Host        string
+	BasePath    string
+	Schemes     []string
+	Title       string
+	Description string
+}
+
+// SwaggerInfo holds exported Swagger Info so clients can modify it
+var SwaggerInfo = swaggerInfo{
```
Comment (rleungx) [inline cmd/pd-server/docs/docs.go:39]: not related to this PR?

Comment (HunDunDM) [reply cmd/pd-server/docs/docs.go:39]: This file should be ignored

Comment (qidi1) [reply cmd/pd-server/docs/docs.go:39]: yes,not related.Maybe some wrong operation caused these changes

**Comment by qidi1** [comment]
> Maybe `/hotspot/history/regions` is more RESTful?

please comment in here https://github.com/tikv/pd/pull/4021

**Comment by IcePigZDB** [comment]
> Maybe `/hotspot/history/regions` is more RESTful?

The style of `/pd/api/v1/hotspot/regions/history`  is more consistent with `HotRead` and `HotWrite`.
```go
HotRead        = "/pd/api/v1/hotspot/regions/read"
HotWrite       = "/pd/api/v1/hotspot/regions/write"
HotHistory     = "/pd/api/v1/hotspot/regions/history"
```

**Pull request closed without merging**